File: notification_service/src/notification/application/dtos/notification.py
from dataclasses import dataclass
from typing import Any, Self

from src.notification.domain.entity import Notification
from src.notification.domain.enums import NotificationStatus


@dataclass
class NotificationSendResult:
    success: bool
    status: NotificationStatus
    external_id: str | None = None


@dataclass
class NotificationContent:
    message: str
    subject: str | None = None


@dataclass
class NotificationDTO:
    """Data Transfer Object for Notification aggregate"""

    id: str
    status: str
    type: str
    destination: str
    template: str
    data: dict[str, Any]
    metadata: dict[str, Any] | None = None
    external_id: str | None = None
    message: str | None = None
    subject: str | None = None

    def add_notification_content(self, content: NotificationContent) -> None:
        self.message = content.message
        self.subject = content.subject

    @classmethod
    def from_domain(cls, notification: Notification) -> Self:
        """Create DTO from domain aggregate"""

        return cls(
            id=str(notification.id_.value),
            status=notification.status,
            type=notification.type,
            destination=notification.destination,
            template=notification.template,
            data=notification.data,
            metadata=notification.metadata,
            external_id=notification.external_id,
        )

    def to_dict(self) -> dict[str, Any]:
        """Convert DTO to dictionary for JSON serialization"""
        return {
            "id": self.id,
            "status": self.status,
            "type": self.type,
            "destination": self.destination,
            "template": self.template,
            "data": self.data,
            "metadata": self.metadata,
            "external_id": self.external_id,
        }

File: rust_notification_service/src/application/dtos/notification.rs
use std::collections::HashMap;

use crate::domain::entity::Notification;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NotificationDTO {
    pub id: String,
    pub status: String,
    pub channel: String,
    pub destination: String,
    pub template: String,
    pub data: HashMap<String, String>,
    pub metadata: Option<HashMap<String, String>>,
    pub external_id: Option<String>,
}

impl From<Notification> for NotificationDTO {
    fn from(notification: Notification) -> Self {
        Self {
            id: notification.id.to_string(),
            status: notification.status.to_string(),
            channel: notification.channel.to_string(),
            destination: notification.destination,
            template: notification.template.to_string(),
            data: notification.data,
            metadata: notification.metadata,
            external_id: notification.external_id,
        }
    }
}
